File: e2e-test/flink-agents-end-to-end-tests-resource-cross-language/src/test/java/org/apache/flink/agents/resource/test/VectorStoreCrossLanguageAgent.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.apache.flink.agents.resource.test;

import org.apache.flink.agents.api.InputEvent;
import org.apache.flink.agents.api.OutputEvent;
import org.apache.flink.agents.api.agents.Agent;
import org.apache.flink.agents.api.annotation.Action;
import org.apache.flink.agents.api.annotation.EmbeddingModelConnection;
import org.apache.flink.agents.api.annotation.EmbeddingModelSetup;
import org.apache.flink.agents.api.annotation.VectorStore;
import org.apache.flink.agents.api.context.MemoryObject;
import org.apache.flink.agents.api.context.RunnerContext;
import org.apache.flink.agents.api.event.ContextRetrievalRequestEvent;
import org.apache.flink.agents.api.event.ContextRetrievalResponseEvent;
import org.apache.flink.agents.api.resource.ResourceDescriptor;
import org.apache.flink.agents.api.resource.ResourceType;
import org.apache.flink.agents.api.vectorstores.CollectionManageableVectorStore;
import org.apache.flink.agents.api.vectorstores.Document;
import org.apache.flink.agents.api.vectorstores.python.PythonCollectionManageableVectorStore;
import org.junit.jupiter.api.Assertions;
import pemja.core.PythonException;

import java.util.HashMap;
import java.util.List;
import java.util.Map;

import static org.apache.flink.agents.api.resource.Constant.OLLAMA_EMBEDDING_MODEL_CONNECTION;
import static org.apache.flink.agents.api.resource.Constant.OLLAMA_EMBEDDING_MODEL_SETUP;
import static org.apache.flink.agents.api.resource.Constant.PYTHON_COLLECTION_MANAGEABLE_VECTOR_STORE;
import static org.apache.flink.agents.api.resource.Constant.PYTHON_EMBEDDING_MODEL_CONNECTION;
import static org.apache.flink.agents.api.resource.Constant.PYTHON_EMBEDDING_MODEL_SETUP;

/**
 * Integration test agent for verifying vector store functionality with Python vector store
 * implementation.
 *
 * <p>This test agent validates: - Python vector store resource registration and creation - Python
 * embedding model integration with vector store - Collection management operations (create, get,
 * delete) - Document CRUD operations (add, get, delete, size) - Context retrieval and similarity
 * search - Cross-language resource dependency (vector store depends on embedding model)
 *
 * <p>Used for e2e testing of the vector store subsystem with cross-language support.
 */
public class VectorStoreCrossLanguageAgent extends Agent {
    public static final String OLLAMA_MODEL = "nomic-embed-text";
    public static final String TEST_COLLECTION = "test_collection";

    @EmbeddingModelConnection
    public static ResourceDescriptor embeddingConnection() {
        if (System.getProperty("EMBEDDING_TYPE", "PYTHON").equals("PYTHON")) {
            return ResourceDescriptor.Builder.newBuilder(PYTHON_EMBEDDING_MODEL_CONNECTION)
                    .addInitialArgument(
                            "module",
                            "flink_agents.integrations.embedding_models.local.ollama_embedding_model")
                    .addInitialArgument("clazz", "OllamaEmbeddingModelConnection")
                    .build();
        } else {
            return ResourceDescriptor.Builder.newBuilder(OLLAMA_EMBEDDING_MODEL_CONNECTION)
                    .addInitialArgument("host", "http://localhost:11434")
                    .addInitialArgument("timeout", 60)
                    .build();
        }
    }

    @EmbeddingModelSetup
    public static ResourceDescriptor embeddingModel() {
        if (System.getProperty("EMBEDDING_TYPE", "PYTHON").equals("PYTHON")) {
            return ResourceDescriptor.Builder.newBuilder(PYTHON_EMBEDDING_MODEL_SETUP)
                    .addInitialArgument(
                            "module",
                            "flink_agents.integrations.embedding_models.local.ollama_embedding_model")
                    .addInitialArgument("clazz", "OllamaEmbeddingModelSetup")
                    .addInitialArgument("connection", "embeddingConnection")
                    .addInitialArgument("model", OLLAMA_MODEL)
                    .build();
        } else {
            return ResourceDescriptor.Builder.newBuilder(OLLAMA_EMBEDDING_MODEL_SETUP)
                    .addInitialArgument("connection", "embeddingConnection")
                    .addInitialArgument("model", OLLAMA_MODEL)
                    .build();
        }
    }

    @VectorStore
    public static ResourceDescriptor vectorStore() {
        return ResourceDescriptor.Builder.newBuilder(PYTHON_COLLECTION_MANAGEABLE_VECTOR_STORE)
                .addInitialArgument(
                        "module",
                        "flink_agents.integrations.vector_stores.chroma.chroma_vector_store")
                .addInitialArgument("clazz", "ChromaVectorStore")
                .addInitialArgument("embedding_model", "embeddingModel")
                .build();
    }

    @Action(listenEvents = InputEvent.class)
    public static void inputEvent(InputEvent event, RunnerContext ctx) throws Exception {
        final String input = (String) event.getInput();

        MemoryObject isInitialized = ctx.getShortTermMemory().get("is_initialized");
        if (isInitialized == null) {
            PythonCollectionManageableVectorStore vectorStore =
                    (PythonCollectionManageableVectorStore)
                            ctx.getResource("vectorStore", ResourceType.VECTOR_STORE);

            // Initialize vector store
            vectorStore.getOrCreateCollection(
                    TEST_COLLECTION, Map.of("key1", "value1", "key2", "value2"));

            CollectionManageableVectorStore.Collection collection =
                    vectorStore.getCollection(TEST_COLLECTION);
            Assertions.assertNotEquals(collection, null, "Vector store collection is null");
            Assertions.assertEquals(
                    TEST_COLLECTION,
                    collection.getName(),
                    "Vector store collection name is not test_collection");
            Assertions.assertEquals(
                    Map.of("key1", "value1", "key2", "value2"),
                    collection.getMetadata(),
                    "Vector store collection metadata is not correct");

            System.out.println("[TEST] Vector store Collection Management PASSED");

            vectorStore.deleteCollection(TEST_COLLECTION);
            Assertions.assertThrows(
                    PythonException.class, () -> vectorStore.getCollection(TEST_COLLECTION));

            // Initialize collection
            vectorStore.add(
                    List.of(
                            new Document(
                                    "Apache Flink Agents is an Agentic AI framework based on Apache Flink.",
                                    Map.of("category", "ai-agent", "source", "test"),
                                    "doc1"),
                            new Document(
                                    "ChromaDB is a vector database for AI applications",
                                    Map.of("category", "database", "source", "test"),
                                    "doc2"),
                            new Document(
                                    "This is a test document used to verify the delete functionality.",
                                    Map.of("category", "utility", "source", "test"),
                                    "doc3")),
                    null,
                    Map.of());

            // Test size
            Assertions.assertEquals(3, vectorStore.size(null), "Vector store size is not 3");

            // Test delete
            vectorStore.delete(List.of("doc3"), null, Map.of());
            Assertions.assertEquals(
                    2, vectorStore.size(null), "Vector store size is not 2, doc3 was not deleted");

            // Test get
            Document doc = vectorStore.get(List.of("doc2"), null, Map.of()).get(0);
            Assertions.assertEquals(
                    "ChromaDB is a vector database for AI applications", doc.getContent());
            Assertions.assertEquals(
                    Map.of("category", "database", "source", "test"), doc.getMetadata());

            System.out.println("[TEST] Vector store Document Management PASSED");

            ctx.getShortTermMemory().set("is_initialized", true);
        }

        ctx.sendEvent(new ContextRetrievalRequestEvent(input, "vectorStore"));
    }

    @Action(listenEvents = ContextRetrievalResponseEvent.class)
    public static void contextRetrievalResponseEvent(
            ContextRetrievalResponseEvent event, RunnerContext ctx) {
        final List<Document> documents = event.getDocuments();

        Map<String, Object> result = new HashMap<>();
        try {
            // Basic validations similar in spirit to EmbeddingIntegrationAgent
            if (documents == null) {
                throw new AssertionError("Vector store returned null documents list");
            }

            if (documents.isEmpty()) {
                throw new AssertionError("Vector store returned empty documents list");
            }

            int idx = 0;
            for (Document doc : documents) {
                if (doc == null) {
                    throw new AssertionError("Document entry is null");
                }

                final String content = doc.getContent();
                if (content == null || content.trim().isEmpty()) {
                    throw new AssertionError(String.format("Document[%d] content is empty", idx));
                }

                // ID can be optional, but when present it must be non-empty
                if (doc.getId() != null && doc.getId().trim().isEmpty()) {
                    throw new AssertionError(String.format("Document[%d] id is empty string", idx));
                }
                idx++;
            }

            result.put("test_status", "PASSED");
            result.put("retrieved_count", documents.size());
            // Include preview of first doc
            Document first = documents.get(0);
            result.put("first_doc_id", first.getId());
            result.put(
                    "first_doc_preview",
                    first.getContent().substring(0, Math.min(50, first.getContent().length())));

            ctx.sendEvent(new OutputEvent(result));
            System.out.printf("[TEST] Vector store retrieval PASSED, count=%d%n", documents.size());
        } catch (Exception e) {
            result.put("test_status", "FAILED");
            result.put("error", e.getMessage());
            ctx.sendEvent(new OutputEvent(result));
            System.err.printf("[TEST] Vector store retrieval FAILED: %s%n", e.getMessage());
            throw e;
        }
    }
}

File: python/flink_agents/e2e_tests/e2e_tests_resource_cross_language/vector_store_cross_language_agent.py
################################################################################
#  Licensed to the Apache Software Foundation (ASF) under one
#  or more contributor license agreements.  See the NOTICE file
#  distributed with this work for additional information
#  regarding copyright ownership.  The ASF licenses this file
#  to you under the Apache License, Version 2.0 (the
#  "License"); you may not use this file except in compliance
#  with the License.  You may obtain a copy of the License at
#
#      http://www.apache.org/licenses/LICENSE-2.0
#
#  Unless required by applicable law or agreed to in writing, software
#  distributed under the License is distributed on an "AS IS" BASIS,
#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
#  See the License for the specific language governing permissions and
# limitations under the License.
#################################################################################
import os
import time

import pytest

from flink_agents.api.agents.agent import Agent
from flink_agents.api.decorators import (
    action,
    embedding_model_connection,
    embedding_model_setup,
    vector_store,
)
from flink_agents.api.events.context_retrieval_event import (
    ContextRetrievalRequestEvent,
    ContextRetrievalResponseEvent,
)
from flink_agents.api.events.event import InputEvent, OutputEvent
from flink_agents.api.resource import Constant, ResourceDescriptor, ResourceType
from flink_agents.api.runner_context import RunnerContext
from flink_agents.api.vector_stores.vector_store import (
    CollectionManageableVectorStore,
    Document,
)

TEST_COLLECTION = "test_collection"
MAX_RETRIES_TIMES = 10

class VectorStoreCrossLanguageAgent(Agent):
    """Example agent demonstrating cross-language embedding model testing."""

    @embedding_model_connection
    @staticmethod
    def embedding_model_connection() -> ResourceDescriptor:
        """EmbeddingModelConnection responsible for ollama model service connection."""
        if os.environ.get("EMBEDDING_TYPE") == "JAVA":
            return ResourceDescriptor(
                clazz=Constant.JAVA_EMBEDDING_MODEL_CONNECTION,
                java_clazz="org.apache.flink.agents.integrations.embeddingmodels.ollama.OllamaEmbeddingModelConnection",
                host="http://localhost:11434",
            )
        return ResourceDescriptor(
            clazz=Constant.OLLAMA_EMBEDDING_MODEL_CONNECTION,
            host="http://localhost:11434",
        )

    @embedding_model_setup
    @staticmethod
    def embedding_model() -> ResourceDescriptor:
        """EmbeddingModel which focus on math, and reuse ChatModelConnection."""
        if os.environ.get("EMBEDDING_TYPE") == "JAVA":
            return ResourceDescriptor(
                clazz=Constant.JAVA_EMBEDDING_MODEL_SETUP,
                java_clazz="org.apache.flink.agents.integrations.embeddingmodels.ollama.OllamaEmbeddingModelSetup",
                connection="embedding_model_connection",
                model=os.environ.get(
                    "OLLAMA_EMBEDDING_MODEL", "nomic-embed-text:latest"
                ),
            )
        return ResourceDescriptor(
            clazz=Constant.OLLAMA_EMBEDDING_MODEL_SETUP,
            connection="embedding_model_connection",
            model=os.environ.get("OLLAMA_EMBEDDING_MODEL", "nomic-embed-text:latest"),
        )

    @vector_store
    @staticmethod
    def vector_store() -> ResourceDescriptor:
        """Vector store setup for knowledge base."""
        return ResourceDescriptor(
            clazz=Constant.JAVA_COLLECTION_MANAGEABLE_VECTOR_STORE,
            java_clazz="org.apache.flink.agents.integrations.vectorstores.elasticsearch.ElasticsearchVectorStore",
            embedding_model="embedding_model",
            host=os.environ.get("ES_HOST"),
            index="my_documents",
            dims=768,
        )

    @action(InputEvent)
    @staticmethod
    def process_input(event: InputEvent, ctx: RunnerContext) -> None:
        """User defined action for processing input.

        In this action, we will test Vector store Collection Management and
        Document Management.
        """
        input_text = event.input

        stm = ctx.short_term_memory

        is_initialized = stm.get("is_initialized") or False

        if not is_initialized:
            print("[TEST] Initializing vector store...")

            vector_store = ctx.get_resource("vector_store", ResourceType.VECTOR_STORE)
            if isinstance(vector_store, CollectionManageableVectorStore):
                vector_store.get_or_create_collection(TEST_COLLECTION , metadata={"key1": "value1", "key2": "value2"})

                collection = vector_store.get_collection(name=TEST_COLLECTION)

                assert collection is not None
                assert collection.name == TEST_COLLECTION
                assert collection.metadata == {"key1": "value1", "key2": "value2"}

                vector_store.delete_collection(name=TEST_COLLECTION)

                with pytest.raises(RuntimeError):
                    vector_store.get_collection(name=TEST_COLLECTION)

                print("[TEST] Vector store Collection Management PASSED")

                documents = [
                    Document(
                        id="doc1",
                        content="The sum of 1 and 2 equals 3.",
                        metadata={"category": "calculate", "source": "test"},
                    ),
                    Document(
                        id="doc2",
                        content="Why did the cat sit on the computer? Because it wanted to keep an eye on the mouse.",
                        metadata={"category": "ai-agent", "source": "test"},
                    ),
                    Document(
                        id="doc3",
                        content="This is a test document used to verify the delete functionality.",
                        metadata={"category": "utility", "source": "test"},
                    ),
                ]
                vector_store.add(documents=documents)

                # Test size
                assert vector_store.size() == 3

                # Test delete
                vector_store.delete(ids="doc3")

                # Wait for vector store to delete doc3
                retry_time = 0
                while vector_store.size() > 2 and retry_time < MAX_RETRIES_TIMES:
                    retry_time += 1
                    time.sleep(2)
                    print(f"[TEST] Retrying to delete doc3, retry_time={retry_time}")

                assert vector_store.size() == 2

                # Test get
                doc = vector_store.get(ids="doc2")
                assert doc is not None
                assert doc[0].id == "doc2"
                assert doc[0].content == "Why did the cat sit on the computer? Because it wanted to keep an eye on the mouse."

                print("[TEST] Vector store Document Management PASSED")

            stm.set("is_initialized", True)


        ctx.send_event(ContextRetrievalRequestEvent(query=input_text, vector_store="vector_store"))

    @action(ContextRetrievalResponseEvent)
    @staticmethod
    def contextRetrievalResponseEvent(event: ContextRetrievalResponseEvent, ctx: RunnerContext) -> None:
        """User defined action for processing context retrieval response.

        In this action, we will test Vector store Context Retrieval.
        """
        documents = event.documents

        assert documents is not None
        assert len(documents) > 0

        for document in documents:
            assert document is not None
            assert document.id is not None
            assert document.content is not None

        test_result = f"[PASS] retrieved_count={len(documents)}, first_doc_id={documents[0].id}, first_doc_preview={documents[0].content[:50]}"
        print(f"[TEST] Vector store Context Retrieval PASSED, first_doc_id={documents[0].id}, first_doc_preview={documents[0].content[:50]}")

        ctx.send_event(OutputEvent(output=test_result))
